File: apps/bookmarks/src/favicon.rs
use scraper::{Html, Selector};
use std::time::Duration;
use url::Url;

fn build_client() -> reqwest::Client {
    reqwest::Client::builder()
        .timeout(Duration::from_secs(15))
        .user_agent("andromeda-bookmarks/0.1 (+https://github.com/stevedylandev/andromeda)")
        .build()
        .expect("Failed to build HTTP client")
}

/// Best-effort favicon URL for a page. Parses `<link rel="icon">` from the
/// HTML, falls back to `/favicon.ico` at the site root. Returns None if
/// the URL is invalid.
pub async fn discover_favicon(page_url: &str) -> Option<String> {
    let parsed = Url::parse(page_url).ok()?;
    let client = build_client();

    if let Ok(resp) = client.get(page_url).send().await {
        if let Ok(body) = resp.text().await {
            let document = Html::parse_document(&body);
            let selector = Selector::parse(
                r#"link[rel="icon"], link[rel="shortcut icon"], link[rel="apple-touch-icon"]"#,
            )
            .ok()?;
            if let Some(href) = document
                .select(&selector)
                .find_map(|el| el.attr("href"))
            {
                if let Ok(resolved) = parsed.join(href) {
                    return Some(resolved.to_string());
                }
            }
        }
    }

    parsed.join("/favicon.ico").ok().map(|u| u.to_string())
}

File: apps/bookmarks-go/favicon.go
package main

import (
	"context"
	"io"
	"net/http"
	"net/url"
	"strings"
	"time"

	"golang.org/x/net/html"
)

const faviconUA = "andromeda-bookmarks-go/0.1 (+https://github.com/stevedylandev/andromeda)"

func discoverFavicon(ctx context.Context, pageURL string) string {
	parsed, err := url.Parse(pageURL)
	if err != nil {
		return ""
	}
	client := &http.Client{Timeout: 15 * time.Second}
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, pageURL, nil)
	if err != nil {
		return ""
	}
	req.Header.Set("User-Agent", faviconUA)
	if resp, err := client.Do(req); err == nil {
		defer resp.Body.Close()
		body, _ := io.ReadAll(io.LimitReader(resp.Body, 1<<20))
		if href := findFaviconHref(string(body)); href != "" {
			if u, err := parsed.Parse(href); err == nil {
				return u.String()
			}
		}
	}
	if u, err := parsed.Parse("/favicon.ico"); err == nil {
		return u.String()
	}
	return ""
}

func findFaviconHref(doc string) string {
	node, err := html.Parse(strings.NewReader(doc))
	if err != nil {
		return ""
	}
	wants := []string{"icon", "shortcut icon", "apple-touch-icon"}
	var found string
	var walk func(*html.Node)
	walk = func(n *html.Node) {
		if found != "" {
			return
		}
		if n.Type == html.ElementNode && strings.EqualFold(n.Data, "link") {
			rel, href := "", ""
			for _, a := range n.Attr {
				switch strings.ToLower(a.Key) {
				case "rel":
					rel = strings.ToLower(strings.TrimSpace(a.Val))
				case "href":
					href = a.Val
				}
			}
			for _, want := range wants {
				if rel == want {
					if href != "" {
						found = href
					}
					return
				}
			}
		}
		for c := n.FirstChild; c != nil; c = c.NextSibling {
			walk(c)
		}
	}
	walk(node)
	return found
}
